rviz_default_plugins/src/rviz_default_plugins/displays/tf/tf_display.cpp

```diff
@@ -287,9 +298,33 @@ void TFDisplay::updateFrames()
 {
   typedef std::vector<std::string> V_string;
   V_string frames = context_->getFrameManager()->getAllFrameNames();
-  std::sort(frames.begin(), frames.end());
 
-  S_FrameInfo current_frames = createOrUpdateFrames(frames);
+  // filter frames according to white-list and black-list regular expressions
+  auto it = frames.begin(), end = frames.end();
+  while (it != end) {
+    if (it->empty() || !std::regex_search(*it, filter_whitelist_property_->regex()) ||
+      std::regex_search(*it, filter_blacklist_property_->regex()))
+    {
+      std::swap(*it, *--end);  // swap current to-be-dropped name with last one
+    } else {
+      ++it;
+    }
+  }
+
+  std::sort(frames.begin(), end);
+
+  S_FrameInfo current_frames;
+  for (it = frames.begin(); it != end; ++it) {
+    FrameInfo * info = getFrameInfo(*it);
+    if (!info) {
+      info = createFrame(*it);
+    } else {
+      updateFrame(info);
+    }
+
+    current_frames.insert(info);
+  }
```

Comment:
It feels to me like these loops are going to be doing an unnecessary amount of work, in both the case where filtering is used and where it is not used.  I'm particularly worried about large TF trees.

For the first loop, if someone is *not* using the filtering, we are still going to be iterating over every single frame to just ignore it.  It seems like if both the whitelist and the blacklist are empty, we should skip the loop completely.

If the user *is* using the filtering, then we are iterating over the list of frames 3 times; once to apply the regex, once to sort, and once to build up the current frames.  It seems to me that we should be able to get away with only iterating once.  That is, we have a loop, and in the loop body we compare it to the whitelist and blacklist.  If it is not to be shown because it is explicitly in the blacklist, or it is not in the whitelist (or it is empty), we do nothing more.  If we are going to show it, then we create/update the frame for it, and insert it into `current_frames`.  Since `current_frames` is a set *anyway* (which is sorted: https://en.cppreference.com/w/cpp/container/set), there is no need for a separate sort step.  We can do a similar thing for the case where the whitelist and blacklist are empty, and skip the additional sort step.

At least, that's how it seems to me.  Does that make sense to you?  Am I possibly missing something with the separate `sort` and "insert into sorted set" step?

Reply:
what do you think @clalancette ? https://github.com/ros2/rviz/pull/1032/commits/246e0d7d4d058b2a2249bd145b0a04a8f7360bfa